content/post/flinksql-time.adoc

```diff
@@ -0,0 +1,1033 @@
+
+* The time the record was written in the system
+* The time a record was first created
+* The time a record was last updated
+* Arbitrary time fields on a record, such as an order record with times that the order was place, fulfilled, and shipped
+
+I'll come back to these in detail later.
+For now, just keep in mind that there is no single "correct" time to use.
+It depends on what your processing is supposed to be doing.
+If you are trying to calculate how many orders were shipped in the last hour, then using the time that the order was created or that it landed in the processing system would give you _an_ answer—it would just be the wrong one.
+
+TIP: https://www.oreilly.com/radar/the-world-beyond-batch-streaming-101/[This seminal article from Tyler Akidau] is nearly ten years old, but is still a highly relevant and excellent read and covers a lot of what you need to know about time in stream processing.
+
+== Time in Kafka
+
+Each Kafka message is made up of https://kafka.apache.org/documentation/#record[several parts]:
+
+. Key
+. Value (the payload)
+. Header
+. Timestamp
+
+Here's an example record.
+The value is in JSON, which I've pretty-printed to make is easier to follow:
+
+[source,javascript]
+----
+Key (14 bytes): {"order_id":1}
+Value (205 bytes):
+    {
+        "order_id": 1,
+        "customer_id": 1001,
+        "total_amount": 149.99,
+        "status": "pending",
+        "created_at": "2025-04-25 09:44:25",
+        "shipped_at": null,
+        "shipping_address": "221B Baker Street, London",
+        "payment_method": "Credit Card"
+    }
+Timestamp: 1745587589625
+----
+
+There are two times here:
+
+. The _event time_ which we'll take as `created_at` from the value: `2025-04-25 09:44:25`.
+. The _processing time_ which is stored in the Kafka message timestamp `1745587589625`.
+This is the milliseconds since the epoch, and https://www.epochconverter.com/[converts] to `2025-04-25 13:26:29.625`.
+
+The timestamp of a Kafka message can be set explicitly https://kafka.apache.org/40/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html[by the producer].
+If it's not set explicitly then the producer sets the timestamp to the current time when it sends the message to the broker.
+The timestamp that's included in the message written by the broker depends on the https://kafka.apache.org/documentation/#topicconfigs_message.timestamp.type[`message.timestamp.type` topic configuration].
+If this is set to `CreateTime` then it uses the time from the producer (either explicit or implicitly set), or it can use the time on the broker when the record is written (`LogAppendTime`).
+
+In theory you could use the Timestamp field of the Kafka message to hold the event time; the data is being stored anyway, so why not optimise by not holding it twice?
+The disadvantage of this comes if you're using a system which doesn't expose the timestamp metadata of a message, so if it doubt, keep it simple :)
```

Comment:
Also there's value in having both of these value: it tells you what's the delay between creation of events and ingesting them into Kafka.

Reply:
good point. added :)